controllers/devworkspace/solver/che_routing.go

```diff
@@ -332,33 +350,115 @@ func (c *CheRoutingSolver) getGatewayConfigsAndFillRoutingObjects(cheCluster *ch
 
 	configs := make([]corev1.ConfigMap, 0)
 
+	useLegacyRouting := false
+	username, err := getNormalizedUsername(c.client, routing.Namespace)
+	if err != nil {
+		useLegacyRouting = true
+	}
+
+	dwName, err := getNormalizedWkspName(c.client, routing.Namespace, routing.Name)
+	if err != nil {
+		useLegacyRouting = true
+	}
+
+	var endpointStrategy EndpointStrategy
+	if useLegacyRouting {
+		endpointStrategy = Legacy{workspaceID: workspaceID}
+	} else {
+		endpointStrategy = UsernameWkspName{username: username, workspaceName: dwName}
+	}
+
 	// first do routing from main che-gateway into workspace service
-	if mainWsRouteConfig, err := provisionMainWorkspaceRoute(cheCluster, routing, cmLabels); err != nil {
+	if mainWsRouteConfig, err := provisionMainWorkspaceRoute(cheCluster, routing, cmLabels, endpointStrategy); err != nil {
 		return nil, err
 	} else {
 		configs = append(configs, *mainWsRouteConfig)
 	}
 
 	// then expose the endpoints
-	if infraExposer, err := c.getInfraSpecificExposer(cheCluster, routing, objs); err != nil {
+	if infraExposer, err := c.getInfraSpecificExposer(cheCluster, routing, objs, endpointStrategy); err != nil {
 		return nil, err
 	} else {
-		if workspaceConfig := exposeAllEndpoints(cheCluster, routing, objs, infraExposer); workspaceConfig != nil {
+		if workspaceConfig := exposeAllEndpoints(cheCluster, routing, objs, infraExposer, endpointStrategy); workspaceConfig != nil {
 			configs = append(configs, *workspaceConfig)
 		}
 	}
 
 	return configs, nil
 }
 
-func (c *CheRoutingSolver) getInfraSpecificExposer(cheCluster *chev2.CheCluster, routing *dwo.DevWorkspaceRouting, objs *solvers.RoutingObjects) (func(info *EndpointInfo), error) {
+func getNormalizedUsername(c client.Client, namespace string) (string, error) {
+	username, err := getUsernameFromNamespace(c, namespace)
```

Comment:
The username is also being read from the namespace as well, in the case that the admin has pre-provisioned the namepsaces following the [doc](https://www.eclipse.org/che/docs/stable/administration-guide/provisioning-namespaces-in-advance/), since in this case, the `user-profile` secret would not be created